### PR TITLE
[InstCombine] Add more cases for simplifying `(icmp (and/or x, Mask), y)`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[InstCombine] Add more tests for simplifying `(icmp (and/or x, Mask), y)`; NFC**
- **[InstCombine] Add more cases for simplifying `(icmp (and/or x, Mask), y)`**

This cleans up basically all the regressions assosiated from #84688

Proof of all new cases: https://alive2.llvm.org/ce/z/5yYWLb

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll [hunk 1]
```diff
@@ -657,3 +675,238 @@ define i1 @src_is_mask_const_sge(i8 %x_in) {
...
+  %r = icmp ult i8 %and, %not_mask
+  ret i1 %r
+}
+
+define i1 @src_x_and_nmask_uge(i8 %x, i8 %y, i1 %cond) {
+; CHECK-LABEL: @src_x_and_nmask_uge(
+; CHECK-NEXT:    [[NOT_MASK0:%.*]] = shl nsw i8 -1, [[Y:%.*]]
+; CHECK-NEXT:    [[R1:%.*]] = icmp ule i8 [[NOT_MASK0]], [[X:%.*]]
+; CHECK-NEXT:    [[NOT_COND:%.*]] = xor i1 [[COND:%.*]], true
+; CHECK-NEXT:    [[R:%.*]] = select i1 [[NOT_COND]], i1 true, i1 [[R1]]
```
Comment (goldsteinn) [inline llvm/test/Transforms/InstCombine/icmp-and-lowbit-mask.ll:793]: NB: We do kind of save an instruction here.
The reason its the same before/after is we run into the case where we add instruction to canonicalize `select` to logical_or

**Comment by goldsteinn** [comment]
ping

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -4189,82 +4191,133 @@ static bool isMaskOrZero(const Value *V, bool Not, const SimplifyQuery &Q,
...
+        return false;
+      };
+      if (Check())
+        return true;
+      std::swap(X, M);
+
+      if (Check())
+        return true;
+
+      return false;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4288]:
```suggestion
      return Check();
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -4189,82 +4191,133 @@ static bool isMaskOrZero(const Value *V, bool Not, const SimplifyQuery &Q,
...
+        }
+        return false;
+      };
+      if (Check())
+        return true;
+      std::swap(X, M);
+      if (Check())
+        return true;
+
+      return false;
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4309]:
```suggestion
      return Check();
```

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 4]
```diff
@@ -4177,7 +4177,9 @@ static bool isMaskOrZero(const Value *V, bool Not, const SimplifyQuery &Q,
 /// a check for a lossy truncation.
 /// Folds:
 ///   icmp SrcPred (x & Mask), x    to    icmp DstPred x, Mask
+///   icmp SrcPred (x & ~Mask), ~Mask    to    icmp DstPred x, ~Mask
 ///   icmp eq/ne (x & ~Mask), 0     to    icmp DstPred x, Mask
+///   icmp eq/ne (x | ~Mask), -1     to    icmp DstPred x, Mask
```
Comment (nikic) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:4182]: Is this supposed to be?
```suggestion
///   icmp eq/ne (~x | Mask), -1     to    icmp DstPred x, Mask
```

**Review by nikic — APPROVED** [review]
LGTM

**Pull request closed without merging**